tests/std/tests/P0660R10_stop_token/test.cpp

```diff
@@ -0,0 +1,418 @@
+    }
+
+    reset_new_counters(1);
+    {
+        stop_source source;
+        auto token = source.get_token();
+        token      = stop_token{}; // rvalue
+        assert(!token.stop_possible());
+    }
+
+    // callback calling in the ctor
+    reset_new_counters(1);
+    {
+        atomic<int> calls{0};
+        stop_source source;
+        source.request_stop();
+        assert(calls.load() == 0);
+        stop_callback<call_counting_functor> cb{source.get_token(), &calls};
+        (void) cb;
+        assert(calls.load() == 1);
+    }
+
+    reset_new_counters(1);
+    {
+        atomic<int> calls{0};
+        stop_token token;
+
+        {
+            stop_source source;
+            token = source.get_token();
+            source.request_stop();
+        } // destroy source
+
+        assert(calls.load() == 0);
+        stop_callback<call_counting_functor> cb{token, &calls};
+        (void) cb;
+        assert(calls.load() == 1);
+    }
+
+    // callback calling on cancel
+    reset_new_counters(1);
+    {
+        atomic<int> calls{0};
+        stop_source source;
+        assert(calls.load() == 0);
+        stop_callback<call_counting_functor> cb{source.get_token(), &calls};
+        assert(calls.load() == 0);
+        source.request_stop();
+        assert(calls.load() == 1);
+    }
+
+    // if the callback is executing on the current thread it does not block for the callback to finish executing
+    reset_new_counters(1);
+    {
+        stop_source source;
+        auto token = source.get_token();
+        optional<stop_callback<cb_destroying_functor>> cb;
+        cb.emplace(token, cb);
+        source.request_stop(); // if we don't do what the standard says, this will deadlock
+    }
+
+    // if the callback is executing on another thread it blocks for the callback to finish executing
+    reset_new_counters(2); // nonstandard assumption that our std::thread allocates exactly once
+    {
+        static constexpr chrono::milliseconds callback_wait_length = 5s;
+        static constexpr chrono::milliseconds request_wait_length  = 500ms;
+        stop_source source;
+        atomic<bool> block_request_stop{false};
+        // block_destroy makes it more likely that the timing assumption above is correct because the timer doesn't
+        // start until we know the worker thread is actively running trying to request_stop
+        atomic<bool> block_destroy{false};
+        std::thread worker{[&] {
+            // run the callbacks in the worker thread
+            block_request_stop.wait(false);
+            block_destroy.store(true);
+            block_destroy.notify_one();
+            assert("request_wait_length TIMING ASSUMPTION" && source.request_stop());
+        }};
+
+
+        auto worker_id = worker.get_id();
+        chrono::steady_clock::time_point started_at;
+        {
+            // timing assumption that the main thread will try to destroy cb within request_wait_length

```

Comment:
Is this going to be an issue for the test harness?

Reply:
I hope not? I'm not sure how else you could test this.